Float docs and defaults.

bfredl opened the pull request:
- [x] `nvim_create_buffer` with `scratch` doesn't work, options in buffer must be set with `aucmd_prepbuf`. (need tests)
- [x] Add overview text for floats.
- [ ] add to vim-diff and places.
- [x] Set some reasonable defaults? `set nonumber` and `set winhl=Normal:PMenu` seems like reasonable defaults that `nvim_open_win()` could do implicitly for floats.

bfredl committed:
docs: floating windows introduction

justinmk commented:
should `WITH_BUFFER` be removed entirely?

bfredl commented:
probably. The terminal code can just use `switch_to_win_for_buf` directly, if it works there.

justinmk commented:
A comment above WITH_BUFFER explaining when `aucmd_prepbuf` should be used, would be ideal.

bfredl committed:
window: simplify logic for entering new float

hkupty commented:
Maybe not the right place to ask this, but wouldn't it make sense to have both `nvim_create_buf` and `nvim_open_win` taking their local options?

So instead of trying to combine`nvim_create_buf(listed, scratch)`, I could say `nvim_create_buf( ..., {listed = false, bufhidden = "wipe"})` and same thing for windows...

It'd save a bunch of sequential `nvim_buf_set_option` and `nvim_win_set_option` calls immediately after (which itself makes no difference, but makes the code conciser and clearer about the intention of respective buffer and windows). 

Just thought about mentioning it since there's a possible overhaul in `nvim_open_win`...

bfredl commented:
@hkupty Rather than `nvim_create_buf` and `nvim_open_win` replicating too much functionality of existing API calls (if `nvim_create_buf` takes options, the next person will ask it to take buffer contents, and so forth), my idea is to add wrapper functions (in lua or C) of common use cases. There should probably be a shared abstraction for preview/signature float next to cursor, also to help different such plugins coexist in some reasonable way.

hkupty commented:
@bfredl That makes sense. Thanks for the response.

bfredl committed:
floats: add NormalFloat highlight and 'nonumber' default

bfredl commented:
I added `FloatWin` highlight group. It sets the default background color of float, and links to `Pmenu` in the builtin highlight definitions. It will still be overriden by `set winhl=Normal:SomethingElse` in the float.

justinmk commented on runtime/doc/syntax.txt:
```diff
@@ -4918,6 +4918,8 @@ TermCursorNC	cursor in an unfocused terminal
 							*hl-VertSplit*
 VertSplit	the column separating vertically split windows
+							*hl-FloatWin*
+FloatWin	Normal text in floating window.
```
I'd suggest `NormalFloat` or something else prefixed with `NormalXX`.

bfredl commented:
Changed to `NormalFloat` as per @justinmk suggestion.

justinmk commented on runtime/doc/api.txt:
```diff
@@ -232,7 +232,45 @@ An example of calling the api from vimscript: >
+
+
+==============================================================================
+Floating windows						   *api-floats*
```
Perhaps `api-floatwin` is a better help tag, to avoid overlap with "floating point numbers".

justinmk commented on runtime/doc/api.txt:
```diff
@@ -232,7 +232,45 @@ An example of calling the api from vimscript: >
+the buffer can be highlight using standard functionality, such as syntax
+highlighting, or |api-highlights|.
+
+Per default, floats will use |hl-FloatWin| as normal highlight, which
```
`hl-NormalFloat`

hkupty commented on runtime/doc/api.txt:
```diff
@@ -232,7 +232,45 @@ An example of calling the api from vimscript: >
+    " optional: change highlight, otherwise Pmenu is used
+    call nvim_win_set_option(win, 'winhl', 'Normal:MyHighlight')
 >
+To close the float, |nvim_close_win()| can be used.
```
isn't it `nvim_win_close`instead?

bfredl committed:
docs: update generated API docs

bfredl commented:
@justinmk @hkupty Thanks, updated to suggestions. 

Travis is weird, it fails on `gcc: error: /home/travis/nvim-deps/usr/lib/libunibilium.so: No such file or directory`. But other PR builds today has succeeded.

bfredl committed:
docs: adjust the generated nvim_open_win docs

The format here is not meant to be ideal, just a backstop until
auto-generation has been fixed.

bfredl commented:
replaced with #9722 due to travis issues
